cmd/indexer/internal/datastore/pq/public_key.go

```diff
@@ -0,0 +1,100 @@
+package pq
+
+import (
+	"context"
+	"time"
+
+	"github.com/Masterminds/squirrel"
+	"github.com/jmoiron/sqlx"
+	"github.com/mailchain/mailchain/cmd/indexer/internal/datastore"
+	"github.com/mailchain/mailchain/crypto"
+	"github.com/mailchain/mailchain/crypto/multikey"
+	"github.com/pkg/errors"
+)
+
+// PublicKeyStore database connection object
+type PublicKeyStore struct {
+	db *sqlx.DB
+}
+
+// PublicKeyStore create new pointer to postgres database
+func NewPublicKeyStore(db *sqlx.DB) (datastore.PublicKeyStore, error) {
+	return &PublicKeyStore{db: db}, nil
+}
+
+type publicKey struct {
+	Address       []byte    `db:"address"`
+	PublicKey     []byte    `db:"public_key"`
+	CreatedAt     time.Time `db:"created_at"`
+	UpdatedAt     time.Time `db:"updated_at"`
+	Protocol      uint8     `db:"protocol"`
+	Network       uint8     `db:"network"`
+	PublicKeyType uint8     `db:"public_key_type"`
+}
+
+func (s PublicKeyStore) PutPublicKey(ctx context.Context, protocol, network string, address []byte, pubKey crypto.PublicKey) error {
+	p, n, err := getProtocolNetworkUint8(protocol, network)
+	if err != nil {
+		return errors.WithStack((err))
+	}
+
+	uPublicKeyType, err := getPublicKeyTypeUint8(pubKey.Kind())
+	if err != nil {
+		return errors.WithStack((err))
+	}
+
+	sql, args, err := squirrel.Update("public_keys").
```

Comment:
Hi @ramkumarvenkat this method requires inserting if it does not exist, or updating the value. It does not need to be a transaction. It could update then check if any rows were affected then do an insert, or use [this approach] (https://stackoverflow.com/questions/1109061/insert-on-duplicate-update-in-postgresql/1109198#1109198) interested to hear your thoughts on this.

Reply:
Thanks @robdefeo yes you're right. I was assuming that `squirrel.UpdateBuilder` does upsert internally, but I was wrong. I have fixed it to do update on `Suffix()`

Reply:
@ramkumarvenkat agreed an integration test would be best for this. One of the big use cases for the `indexer` is to create integration tests for Mailchain which will be exciting 😄

Reply:
@ramkumarvenkat good spot with `Suffix()` I was not aware that `squirrel` support that.